src/java.net.http/share/classes/jdk/internal/net/http/HttpClientImpl.java

```diff
@@ -167,7 +191,129 @@ private void shutdown() {
                         new RuntimePermission("modifyThread"));
             }
         }
+    }
+
+    // We maintain a list of pending requests that will be aborted if ever
+    // the selector manager thread exists abnormally.
+    // The request, its id, and its completable future, are stored in a record-like
+    // PendingRequest object added to the pending requests set (pendingRequests).
```

Comment:
I'm not too familiar with the HttpClient code yet. Before this change, was there ever a place where we would end up holding onto the request instances in memory and prevent them from being garbage collected? Would this change cause unintentional resource constraints in a regular/good scenario (where the `SelectorManager` thread hasn't exited abnormally and is still alive) or in cases where HTTP responses are slow to come back from servers?

Reply:
In the normal course of action the HttpClient will not shutdown if any operation are still in progress. So holding onto the multiexchange here doesn't prevent the client from shutting down, since the multiexchange will be remove from this list when the operation completes.